### ISSUE TITLE
root on symlinked path causes panic when using "Watch for changes"

### ISSUE THREAD

**Issue by larroy** [syncthing/syncthing#5043]
I think the logic here has problems when the path is symlinked:

https://github.com/syncthing/syncthing/blob/master/lib/fs/basicfs_watch.go#L113

I have a "storage" folder symlinked to mnt:  /home/piotr/storage -> /mnt/storage/storage/

This causes panic when using the notify backend

panic: bug: Notify backend is processing a change outside of the filesystem root: root==/home/piotr/storage/synced/oneplus5/, rootSymEval==/home/piotr/storage/synced/oneplus5/, path==/mnt/storage/storage/synced/oneplus5/.stfolder

### Version Information

Syncthing Version: v0.14.48
OS Version: Linux 64  Ubuntu 16.04 (installed from apt)
Browser Version: (if applicable, for GUI issues)

**Comment by imsodin** [comment]
Can you run the following script on your system and report the results, i.e. install Go and run `go run test.go`

```golang
package main

import (
	"fmt"
	"path/filepath"
)

func main() {
	fmt.Println(filepath.EvalSymlinks("/home/piotr/storage/synced/oneplus5/"))
}
```

In any case we need to act on, not just ignore an error when determining the root with symlinks evaluated: https://github.com/syncthing/syncthing/blob/master/lib/fs/basicfs.go#L56  
Either panic already there (probably a bad idea, though I don't get why it should ever fail), or disable watching to prevent panics.

**Comment by larroy** [comment]
```
/mnt/storage/storage/synced/oneplus5 <nil>
```

**Comment by larroy** [comment]
Why the assertion in the first place? what are we trying to accomplish in unrootedChecked ?

**Comment by imsodin** [comment]
@larroy 

Why asserting/panicing is a good question - it makes life easy for me as I don't need to think about how to handle a case that should never happen. A panic is ideal for that, as one line of code guarantees that the issue will come to my attention if it ever arises. Obviously, it's not quite user-friendly. All of that is tangential to your actual case though.

What needs figuring out is where the evaluation of symlinks in your filesystem instance failed, i.e. why there is this conflict:

panic: bug: Notify backend is processing a change outside of the filesystem root: root==/home/piotr/storage/synced/oneplus5/, **rootSymEval==/home/piotr/storage/synced/oneplus5/**, path==/mnt/storage/storage/synced/oneplus5/.stfolder

fmt.Println(filepath.EvalSymlinks("/home/piotr/storage/synced/oneplus5/"))
**/mnt/storage/storage/synced/oneplus5** <nil>

I can't reproduce this issue, so a bit of back and forth will be necessary (maybe I am also just missing the obvious). Can you check what's the path for your folder in config.xml and then adjust the path at the bottom of the following gist and run it (it's basically the relevant syncthing code with lots of verbosity):
https://gist.github.com/imsodin/b8ec2ec226a0575f7619997b287e6e28

**Comment by larroy** [comment]
Because the mount point is in /mnt/storage but in my home folder there's a link from ~/storage to /mnt/storage

**Comment by imsodin** [comment]
The point of `rootSymEval` is exactly to resolve this issue, i.e. it should be the evaluated symlink as returned by `filepath.EvalSymlinks` und thus the same. Please run the gist and report the results.

**Comment by jamiesonbecker** [comment]
Same panic happening here, setup similar to larroy's:

/oldssd/jamieson/  $ ls -l /home/
lrwxrwxrwx  1 root     root       17 Feb  6 21:36 jamieson -> /oldssd/jamieson/

Basically ran out of space on one of my SSD's, so just symlinked from another one with free space available. Reproduce: just add a new user and then symlink the home directory from somewhere else.

**Comment by larroy** [comment]
Sorry I haven't been able to rerun, hadn't had access to that desktop for a bit.

**Comment by jamiesonbecker** [comment]
@imsodin sure:

```
Panic at 2018-08-06T15:47:05-05:00
panic: bug: Notify backend is processing a change outside of the filesystem root: 
  root==/home/jamieson/Essential-Ph1-Camera/, 
  rootSymEval==/home/jamieson/Essential-Ph1-Camera/, 
  path==/oldssd/jamieson/Essential-Ph1-Camera/.stfolder
```

**Comment by imsodin** [comment]
The weird thing is, that the gist (which is the exact same code with some debug statements) produces the correct result for the evaluated path `&{/home/jamieson/Essential-Ph1-Camera/ /oldssd/jamieson/Essential-Ph1-Camera/}` but it's wrong in the panic `rootSymEval==/home/jamieson/Essential-Ph1-Camera/,`. I have no idea, the next thing I'd try is move the debug info into the Syncthing binary and test again. A version on top of master from the build servers is here  
https://build.syncthing.net/repository/download/Syncthing_BuildLinuxCross/24994:id/syncthing-linux-amd64-v0.14.50-rc.1%2B3-gf9798d8.tar.gz%21/syncthing-linux-amd64-v0.14.50-rc.1%2B3-gf9798d8/syncthing  
and a version on top of 0.14.49 compile by me is here  
https://polybox.ethz.ch/index.php/s/ImkcFgiqQHxkCHt

Please run either of those until you get the panic again and post the logs here.

**Comment by jamiesonbecker** [comment]
Hmm, I stopped getting the panic once I disabled Auto-Accept and created that folder manually (with the same id).

**Comment by jamiesonbecker** [comment]
(403 on build.syncthing URL)

**Comment by jamiesonbecker** [comment]
If I pause the folder that I'm getting running folder error on (from the same device) instead of disabling Auto-Accept, then I get this nice little issue (this is still running stock v0.14.49 on Arch):

```
[SGBHH] 09:52:17 INFO: Auto-accepted abcdef1234 folder "Ph1" (abc123) at path ~/Ph1
[SGBHH] 09:52:17 INFO: Connection to XXX at [yyy]:58648-[zzz]:22000/tcp-client closed: <nil>
panic: runtime error: invalid memory address or nil pointer dereference
```

(I can email you the full uncensored panic if you like.)

Just a hunch: these two issues (#5063 and this one, #5043) might be related to each other, especially when symlinks are thrown into the mix.

**Comment by imsodin** [comment]
@jamiesonbecker Yes, that hash is correct.  

Maybe you need to login somehow to teamcity before the link can be accessed.

**Comment by jamiesonbecker** [comment]
Your build produces the same nil pointer reference:

```
Panic at 2018-08-07T10:24:04-05:00
panic: runtime error: invalid memory address or nil pointer dereference
    panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x91f1cc]

goroutine 577 [running]:
github.com/syncthing/syncthing/lib/model.(*Model).Closed(0xc420211320, 0x7f3481835e60, 0xc4224d3710, 0x0, 0x0)
    /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/model.go:1262 +0x50c
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).close.func1()
    /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/protocol/protocol.go:833 +0x206
sync.(*Once).Do(0xc420730148, 0xc420efb120)
    /media/ext4_data/Linux/source/go/src/sync/once.go:44 +0xbe
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).close(0xc4207300c0, 0x0, 0x0)
    /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/protocol/protocol.go:820 +0x6b
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).readerLoop.func1(0xc4207300c0, 0xc420efbfd0)
    /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/protocol/protocol.go:335 +0x41
panic(0xa75380, 0x1117ca0)
    /media/ext4_data/Linux/source/go/src/runtime/panic.go:502 +0x229
github.com/syncthing/syncthing/lib/db.(*FileSet).IndexID(0x0, 0xffffffffffffffff, 0xffffffffffffffff, 0xffffffffffffffff, 0xffffffffffffffff, 0xc422e5c501)
    /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/db/set.go:263 +0x5e
github.com/syncthing/syncthing/lib/model.(*Model).ClusterConfig(0xc420211320, 0x4b91a04c26d701c3, 0xeb45ab20275c2e1e, 0x81bc97fd845af037, 0x4df64d3c0e0a8595, 0xc420ea8500, 0x4, 0x4)
    /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/model/model.go:949 +0xda6
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).readerLoop(0xc4207300c0, 0x0, 0x0)
    /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/protocol/protocol.go:361 +0xed0
created by github.com/syncthing/syncthing/lib/protocol.(*rawConnection).Start
    /media/ext4_data/Coding/go/src/github.com/syncthing/syncthing/lib/protocol/protocol.go:225 +0x43

[snip]
```

I'll see if I can reproduce the original symlink related error above.

**Comment by jamiesonbecker** [comment]
This is only occurring when interacting with folders from v.14.49-dirty.

**Comment by jamiesonbecker** [comment]
Resuming the problematic device and folder:

```
[SGBHH] 10:28:35 INFO: newBasicFilesystem windows root hack: /home/jamieson/XXX
[SGBHH] 10:28:35 INFO: newBasicFilesystem tilde expansion: /home/jamieson/XXX
[SGBHH] 10:28:35 INFO: newBasicFilesystem EvalSymlinks: /oldssd/jamieson/XXX
[SGBHH] 10:28:35 INFO: newBasicFilesystem start: ~/Ph1
[SGBHH] 10:28:35 INFO: newBasicFilesystem windows root hack: ~/Ph1
[SGBHH] 10:28:35 INFO: newBasicFilesystem tilde expansion: /home/jamieson/Ph1
[SGBHH] 10:28:35 INFO: newBasicFilesystem EvalSymlinks failed: /home/jamieson/Ph1 lstat /oldssd/jamieson/Ph1: no such file or directory
[SGBHH] 10:28:35 INFO: Adding folder "XXX" (xxx)
panic: cannot start already running folder xxx
[monitor] 10:28:35 WARNING: Panic detected, writing to "zzz/panic-20180807-102835.log"
```

**Comment by jamiesonbecker** [comment]
N.B., this running folder issue is a different folder than the original camera folder shared from device; disabling auto accept got that original folder working, and I haven't been able to reproduce that issue until this new one popped up.

**Comment by imsodin** [comment]
The nil deref is completely unrelated to everything else - thanks anyway for finding it ;)

The cannot start already running folder almost certainly is #5063.

Does a folder for `/oldssd/jamieson/Ph1` actually exist? Because above in the gist and panic, it was `/home/jamieson/Essential-Ph1-Camera/`.

**Comment by jamiesonbecker** [comment]
Great. Yes, that one was fixed as soon as I turned off auto-accept and restarted on both sides. `Ph1` is a new one to try to force reproduction of this issue again, but I haven't gotten it to trigger the same symlink issue. (The directory wasn't created on the desktop side, so it still doesn't exist.)

**Comment by imsodin** [comment]
@jamiesonbecker 

I still don't understand what exactly is happening, but regardless the way symlinks in the root path are handled today is clearly bad and may lead to panics. Therefore I proposed a change that probably will fix this issue: #5105

The nil deref trace from https://github.com/syncthing/syncthing/issues/5043#issuecomment-411097106 turned out to be less obvious than I thought initially, it's actually very confusing. If you do have the full trace and logs leading to it, I would be very interested to get them (by email or in the forum please, as it's not related to this issue).
